### DISCUSSION TITLE
Conditional turn restrictions

### DISCUSSION THREAD

**Discussion by alexey-gl** [valhalla/valhalla#4539] [Q&A] [answered]
Hello!
I'm  looking for ability to use conditional turn restrictions during routing (dependent on time and costing options attributes).
As well as I see, current turn restrictions implementation simply analyze restriction existence on predcessor without paying attention to any attributes of the restriction.
https://github.com/valhalla/valhalla/blob/master/src/sif/autocost.cc#L424

I have my own Costing class and ideally I would like to make a comprehensive analysis of turn restriction during costing Allowed process.
I have extended DynamicCost::Allowed method for my custom conditions in other ways.
But for turn restrictions Allowed method have only EdgeLabel& pred to access predcessor information, so I could't have access to full information about previous edge.
Please give me some point to find a way of my task implemention.

- Does DynamicCost::Restricted method could help me with my problem? Is it good idea to override it in my custom Costing class?
- If I will try to access restrictions from Allowed method, is it a good idea to get predcessor DirectedEdge info from the tile of current edge in DynamicCost::Allowed method? If it is possible that predcessor stored in another tile, how to get predcessor DirectedEdge info for Costing?

**Comment by nilsnolde** [comment, marked as answer]
Sounds like you should dive deeper into `EvaluateRestrictions` exactly where you are in the code. It should have all you need. We do process time restricted restrictions for many of them (not all). Look at the graph builder, particularly follow the code paths in pbfgraphparser.cc. Or wait for someone else to answer. Or contact me on enquiry@gis-ops.com for commercial support.

**Reply by alexey-gl** [reply]
Thanks a lot for fast answer! Looks like I am on the right way. I will continue my research..

**Comment by alexey-gl** [comment]
@nilsnolde After some time with code analysis I've found that  `EvaluateRestrictions` method check simple restrictions and `Restricted` method analyse complex restrictions (i.e. from A to C via B and so on).
Because I need to customise conditions for such complex restrictions so I think I need to override `Restricted` method in my `Costing` class. Is it right?

One doubt about `Restricted` method is that it looks like it contains too much "algorithm logic" (i.e. change status of edges which have already been evaluated). Looks like this part is not `Costing` class method logic. What do you think about it?
